// @kentico/kontent-delivery/_commonjs/models/content-type/responses.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var TypeResponses;
(function (TypeResponses) {
    var ListContentTypesResponse = /** @class */ (function () {
        /**
        * Response containing multiple types
        * @constructor
        * @param {IContentType[]} types - Content types
        * @param {Pagination} pagination - Pagination object
        */
        function ListContentTypesResponse(types, pagination, debug) {
            this.types = types;
            this.pagination = pagination;
            this.debug = debug;
        }
        return ListContentTypesResponse;
    }());
    TypeResponses.ListContentTypesResponse = ListContentTypesResponse;
    var ViewContentTypeResponse = /** @class */ (function () {
        /**
        * Response containing single type
        * @constructor
        * @param {IContentType} type - Content type
        */
        function ViewContentTypeResponse(type, debug) {
            this.type = type;
            this.debug = debug;
        }
        return ViewContentTypeResponse;
    }());
    TypeResponses.ViewContentTypeResponse = ViewContentTypeResponse;
})(TypeResponses = exports.TypeResponses || (exports.TypeResponses = {}));
