// @kentico/kontent-core/_es2015/http/test-http.service.js
import { of, throwError } from 'rxjs';
var TestHttpService = /** @class */ (function () {
    function TestHttpService(config) {
        this.throwError = false;
        this.fakeResponseJson = undefined;
        this.errorJson = undefined;
        Object.assign(this, config);
    }
    TestHttpService.prototype.retryPromise = function (promise, options) {
        console.log('Retry is not implemented in test service. Returning original Promise');
        return promise;
    };
    TestHttpService.prototype.get = function (call, options) {
        // throw kontent error
        if (this.throwError) {
            var fakeError = {
                response: {
                    data: this.errorJson
                }
            };
            return throwError({
                originalError: fakeError,
                mappedError: call.mapError(fakeError)
            });
        }
        // return fake response
        return of({
            data: this.fakeResponseJson,
            response: undefined
        });
    };
    TestHttpService.prototype.post = function (call, options) {
        // throw kontent error
        if (this.throwError) {
            var fakeError = {
                response: {
                    data: this.errorJson
                }
            };
            return throwError({
                originalError: fakeError,
                mappedError: call.mapError(fakeError)
            });
        }
        // return fake response
        return of({
            data: this.fakeResponseJson,
            response: undefined
        });
    };
    TestHttpService.prototype.put = function (call, options) {
        // throw kontent error
        if (this.throwError) {
            var fakeError = {
                response: {
                    data: this.errorJson
                }
            };
            return throwError({
                originalError: fakeError,
                mappedError: call.mapError(fakeError)
            });
        }
        // return fake response
        return of({
            data: this.fakeResponseJson,
            response: undefined
        });
    };
    TestHttpService.prototype.delete = function (call, options) {
        // throw kontent error
        if (this.throwError) {
            var fakeError = {
                response: {
                    data: this.errorJson
                }
            };
            return throwError({
                originalError: fakeError,
                mappedError: call.mapError(fakeError)
            });
        }
        // return fake response
        return of({
            data: this.fakeResponseJson,
            response: undefined
        });
    };
    return TestHttpService;
}());
export { TestHttpService };
