// @kentico/kontent-delivery/_es2015/models/item/item-models.js
var ContentItem = /** @class */ (function () {
    /**
    * Base class representing content item type. All content type models need to extend this class.
    * @constructor
    */
    function ContentItem(config) {
        this._config = config;
    }
    return ContentItem;
}());
export { ContentItem };
var ContentItemSystemAttributes = /** @class */ (function () {
    function ContentItemSystemAttributes(data) {
        Object.assign(this, data);
    }
    return ContentItemSystemAttributes;
}());
export { ContentItemSystemAttributes };
var Link = /** @class */ (function () {
    function Link(data) {
        Object.assign(this, data);
    }
    return Link;
}());
export { Link };
var RichTextImage = /** @class */ (function () {
    function RichTextImage(data) {
        Object.assign(this, data);
    }
    return RichTextImage;
}());
export { RichTextImage };
export var RichTextItemDataType;
(function (RichTextItemDataType) {
    RichTextItemDataType["Item"] = "item";
})(RichTextItemDataType || (RichTextItemDataType = {}));
